### DISCUSSION TITLE
Sync non-zero exit on fix

### DISCUSSION THREAD

**Discussion by enadeau** [DavidVujic/python-polylith#277] [Ideas]
**Is your feature request related to a problem? Please describe.**
There is no way to tell if the sync command  made some changes or not. This makes it hard to run the command in CI and fail the job if the project is not in sync.

**Describe the solution you'd like**
Add a flag like ` --exit-non-zero-on-fix` to the sync command to fail with non zero exit code if some changed occurred while running the command.

**Describe alternatives you've considered**
The check commands seems like the command that should be used for that but it does not complain if I remove one of the component from the main pyproject file

**Comment by DavidVujic** [comment]
I didn't know that the `poly sync` command was used in CI! My initial idea was to run it locally, and then commit the changes.

I'm curious about why the `poly check` doesn't seem to identify the changes, that sounds like a bug. Or is it the top-most pyproject.toml file you mean? If so, then I understand. The check command does the verifying on the projects only, to be able to switch between implementations/components in development.

**Comment by enadeau** [comment]
The idea to run it in CI was to make sure nobody in the team forgets about it and thing falling out of sync.

Yes I was actually testing it on the root pyproject so it is probably the reason why it didn't catched it. If there was a way for check to catch that to that would also be a good way to fix the issue

**Comment by DavidVujic** [comment]
That makes sense! I have to think about this and how to solve it, because I also believe some teams (during development/feature branches) exclude some components and I don't want that to break the builds. It's not strictly a requirement to include all code in development, even if it is highly recommended.

There is the `--verbose` option, but that is mainly for local development too. Unfortunately I have annotated it incorrectly for that specific command, it should be a boolean flag and not a string. I will make an update to make it behave the same as the same flag for other commands.

**Comment by DavidVujic** [comment]
Would a pre-commit hook be a solution?

**Comment by enadeau** [comment]
I think a pre-commit hook would be a great solution :smiley: 

 I was actually thinking of this flag as some per-requisite for building a hook

**Comment by DavidVujic** [comment]
Would this simple pre-commit do what's needed?

```
#!/usr/bin/env bash
set -eo pipefail

uv run poly sync   # or, if using poetry: poetry poly sync
```

(I created a file named `pre-commit` in the `.git/hooks` folder)

There's of course also the `pre-commit` tool/library with a lot more features 😄

**Comment by enadeau** [comment]
I was more thinking about the pre-commit library when I mentioned that as I'm looking for a reliable, consistent way to deploy it accross the team (or in CI)

**Comment by DavidVujic** [comment]
I think it is possible without a dedicated published hook, by using the already installed poly tool. But I am not that experienced in how the `pre-commit` tool is meant to be used.

I experimented with the `pre-commit` library:

```shell
uv add pre-commit

touch .pre-commit-config.yaml
```

and added this content to the yaml:
```yaml
repos:
-   repo: local
    hooks:
    -   id: poly-sync
        name: run poly sync
        entry: bash -c "uv run poly sync"
        pass_filenames: false
        language: system
```

As an alternative, you could also have a `hook.sh` file with the contents I previously shared, and set the `entry` to the hook instead.

What do you think about that?

(EDIT: this will probably be a problem for Windows users, when relying on bash)

EDIT 2: here's a 🤪 inline Python equivalent 
`python -c "import subprocess; subprocess.run(['uv', 'run', 'poly', 'sync'])"`

**Comment by DavidVujic** [comment]
(moving this to the ideas section in discussions)

**Comment by enadeau** [comment]
This is actually a great solution. I add no idea it was possible to use "local" as a repo. I'll try to implement that for our repo
